src/adapter/graphql/src/queries/datasets/dataset_flow_runs.rs

```diff
@@ -62,20 +63,25 @@ impl DatasetFlowRuns {
         let page = page.unwrap_or(0);
         let per_page = per_page.unwrap_or(Self::DEFAULT_PER_PAGE);
 
-        // TODO: consider pushing pagination control down to service/repository levels
-        let flows_stream = flow_service
-            .list_all_flows_by_dataset(&self.dataset_handle.id)
+        let flows_state_listing = flow_service
+            .list_all_flows_by_dataset(
+                &self.dataset_handle.id,
+                match filters {
+                    Some(filters) => filters.into(),
+                    None => Default::default(),
+                },
+                fs::FlowPaginationOpts {
+                    offset: page * per_page,
+                    limit: per_page,
+                },
+            )
+            .await
             .int_err()?;
 
-        let all_flows: Vec<_> = flows_stream.try_collect().await?;
-        let total_count = all_flows.len();
+        let matched_flows: Vec<_> = flows_state_listing.matched_stream.try_collect().await?;
+        let total_count = flows_state_listing.total_count;
 
-        let nodes = all_flows
-            .into_iter()
-            .skip(page * per_page)
-            .take(per_page)
-            .map(Flow::new)
-            .collect();
+        let nodes = matched_flows.into_iter().map(Flow::new).collect();
```

Comment:
if you move `map()` before `matched_stream.try_collect()` you can avoid allocating one extra `Vec`

Reply:
Corrected